plasmapy/analysis/time_series/conditional_averaging.py

```diff
@@ -0,0 +1,402 @@
+"""Functionality to calculate the conditional average and conditional variance of a time series."""
+
+__all__ = ["ConditionalEvents"]
+
+
+import astropy.units as u
+import numpy as np
+
+from astropy.units import UnitsError
+from scipy.signal import find_peaks
+
+
+class ConditionalEvents:
+    """
+    Calculate conditional average, conditional variance, peaks,
+    arrival times and waiting times of events of a time series.
+
+    Parameters
+    ----------
+    signal : 1D |array_like|
+        Signal to be analyzed.
+    time : 1D |array_like|
+        Corresponding time values for ``signal``.
+    lower_threshold : `float` or `~astropy.units.Quantity`
+        Lower threshold for event detection.
+    upper_threshold : `float` or `~astropy.units.Quantity`, default: `None`
+        Upper threshold for event detection.
+    reference_signal : 1D |array_like|, default: `None`
+        Reference signal.
+        If `None`, ``signal`` is the reference signal.
+    length_of_return : float, default: `None`
+        Desired length of returned data.
+        If `None`, estimated as ``len(signal) / len(number_of_events) * time_step``.
+    distance : float, default: ``0``
+        Minimum distance between peaks, in units of time.
+    remove_non_max_peaks : bool, default: `False`
+        Remove events where peak is not the largest value inside window.
+
+    Raises
+    ------
+    `ValueError`:
+        If length of ``signal`` and ``time`` are not equal.
+        If length of ``reference_signal`` and ``time`` are not equal (when reference_signal is provided).
+        If ``length_of_return`` is greater than the length of the time span.
+        If ``length_of_return`` is negative.
+        If ``upper_threshold`` is less than or equal to ``lower_threshold``.
+
+    `UnitsError`:
+        If astropy units of ``signal``/``reference_signal`` and ``lower_threshold`` do not match.
+        If astropy units of ``signal``/``reference_signal`` and ``upper_threshold`` do not match.
+        If astropy units of ``time``, ``length_of_return`` and ``distance`` do not match.
+
+    Notes
+    -----
+    A detailed analysis of the conditional averaging method is presented in
+    Rolf Nilsen's master thesis: "Conditional averaging of overlapping pulses"
+    https://munin.uit.no/handle/10037/29416
```

Comment:
A small section overviewing the basics of conditional averaging would be instrumental here, also could you change the URL to be a linked reference?